pkg/azurefile/azurefile.go

```diff
@@ -872,6 +883,60 @@ func (d *Driver) ResizeFileShare(ctx context.Context, subsID, resourceGroup, acc
 	})
 }
 
+// CopyFileShare copies a fileshare in the same storage account
+func (d *Driver) CopyFileShare(ctx context.Context, req *csi.CreateVolumeRequest, accountKey string, shareOptions *fileclient.ShareOptions, storageEndpointSuffix string) error {
+	if shareOptions.Protocol == storage.EnabledProtocolsNFS {
+		return fmt.Errorf("protocol nfs is not supported for volume cloning")
+	}
+	var volumeID string
+	if req.GetVolumeContentSource() != nil && req.GetVolumeContentSource().GetVolume() != nil {
+		volumeID = req.GetVolumeContentSource().GetVolume().GetVolumeId()
+	}
+	resourceGroupName, accountName, srcFileShareName, _, _, subsID, err := GetFileShareInfo(volumeID)
+	dstFileShareName := shareOptions.Name
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	klog.V(2).Infof("begin to copy fileshare %s to %s", srcFileShareName, dstFileShareName)
+	if srcFileShareName == "" {
+		return fmt.Errorf("srcFileShareName is empty")
+	}
+	if dstFileShareName == "" {
+		return fmt.Errorf("dstFileShareName is empty")
+	}
+
+	klog.V(2).Infof("generate sas token for account(%s)", accountName)
+	accountSasToken, genErr := generateSASToken(accountName, accountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes)
+	if genErr != nil {
+		return genErr
+	}
+
+	if _, rerr := d.cloud.GetFileShare(ctx, subsID, resourceGroupName, accountName, srcFileShareName); rerr != nil {
+		return fmt.Errorf("srcFileShareName(%s) get error: %s", srcFileShareName, rerr.Error())
+	}
+
+	timeAfter := time.After(waitForCopyTimeout)
+	timeTick := time.Tick(waitForCopyInterval)
+	srcPath := fmt.Sprintf("https://%s.file.%s/%s%s", accountName, storageEndpointSuffix, srcFileShareName, accountSasToken)
+	dstPath := fmt.Sprintf("https://%s.file.%s/%s%s", accountName, storageEndpointSuffix, dstFileShareName, accountSasToken)
+
+	for {
+		select {
+		case <-timeTick:
+			klog.V(2).Infof("copy fileshare %s to %s", srcFileShareName, dstFileShareName)
```

Comment:
there is 5s delay between `generate sas token for account` and `begin to copy file share`

```
I0817 02:46:07.825044       1 azurefile.go:908] generate sas token for account(f0b87b811e4c140479ff7d5)
I0817 02:46:12.950351       1 azurefile.go:926] copy fileshare pvc-86149bb5-6848-4e3e-a190-98a167d0c0e5 to pvc-31811096-5d02-4d54-aaca-db56cd11a0be
I0817 02:46:15.135361       1 azurefile.go:931] copied fileshare pvc-86149bb5-6848-4e3e-a190-98a167d0c0e5 to pvc-31811096-5d02-4d54-aaca-db56cd11a0be successfully
I0817 02:46:15.135435       1 controllerserver.go:556] create file share pvc-31811096-5d02-4d54-aaca-db56cd11a0be on storage account f0b87b811e4c140479ff7d5 successfully
```